app/helpers/payouts_helper.rb

```diff
@@ -156,6 +230,213 @@ def displayable_payout_period_range(previous_payment, payout_period_end_date)
+    payout_period_data[:payout_displayed_amount] = formatted_dollar_amount(payout_period_data[:payout_cents])
+    payout_period_data[:payout_date_formatted] = formatted_payout_date(user.next_payout_date)
+    payout_period_data[:type] = if user.payout_frequency == User::DAILY && Payouts.is_user_payable(user, payout_period_end_date, payout_type: Payouts::PAYOUT_TYPE_INSTANT)
+      Payouts::PAYOUT_TYPE_INSTANT
+    else
+      Payouts::PAYOUT_TYPE_STANDARD
+    end
+  end
+
+  def calculate_carried_over_amounts(user, balance_ids, payout_period_data)
+    carried_over_balance_ids = []
+
+    if balance_ids.present?
+      earlier_failed_list = get_earlier_failed_payments(user, balance_ids, user.next_payout_date)
+
+      if earlier_failed_list.any?
+        latest_failed = earlier_failed_list.max_by(&:created_at)
+        payout_period_data[:carried_over_from_failed_on] = formatted_payout_date(latest_failed.created_at)
+
+        items = []
+        total_cents = 0
+        processed_balance_ids = Set.new
+
+        earlier_failed_list.each do |failed_p|
+          failed_ids = failed_p.balances.pluck(:id)
+          intersect_ids = balance_ids & failed_ids
+          next if intersect_ids.empty?
+
+          # Only process balances that haven't been counted yet
+          new_balance_ids = intersect_ids - processed_balance_ids.to_a
+          next if new_balance_ids.empty?
+
+          carried_over_balance_ids |= new_balance_ids
+          processed_balance_ids.merge(new_balance_ids)
+
+          cents = Balance.where(id: new_balance_ids).sum(:holding_amount_cents)
+          next if cents.zero?
+          total_cents += cents
+          items << {
+            failed_on: formatted_payout_date(failed_p.created_at),
+            carried_over_cents: cents,
+            carried_over_displayed_amount: formatted_dollar_amount(cents)
+          }
+        end
+
+        if total_cents > 0
+          payout_period_data[:carried_over_cents] = total_cents
+          payout_period_data[:carried_over_displayed_amount] = formatted_dollar_amount(total_cents)
+          payout_period_data[:carried_over_items] = items
+        end
+      end
+    end
+
+    carried_over_balance_ids
+  end
+
+  def get_earlier_failed_payments(user, balance_ids, cutoff_date)
+    user.payments
+      .failed
+      .joins(:balances)
+      .where("payments.created_at < ?", cutoff_date)
+      .where(balances: { id: balance_ids })
+      .order("payments.created_at ASC")
+      .distinct
+      .to_a
+  end
+
+  def add_payout_note(payout_period_data, user)
+    last_payout_note = user.comments.with_type_payout_note.where(author_id: GUMROAD_ADMIN_ID).where.not("content like 'Payout via PayPal%'").last
+    payout_period_data[:payout_note] = \
+      if last_payout_note.present? && last_payout_note.created_at.to_i > user.payments.completed_or_processing.last&.created_at.to_i
+        last_payout_note.content.gsub("via Stripe ", "")
```

Comment:
Not really sure why this is needed. Can you please add some explanatory comments in the PR to explain your approach and changes?